lib/types/addresstrie.go

```diff
@@ -0,0 +1,120 @@
+package types
+
+import (
+	"bytes"
+	"encoding/json"
+	"net"
+	"strconv"
+	"strings"
+)
+
+// NullAddressTrie is a wrapper around AddressTrie like guregu/null
+type NullAddressTrie struct {
+	Trie  *AddressTrie
+	Valid bool
+}
+
+// NewNullAddressTrie returns valid (Valid: true) AddressTrie
+func NewNullAddressTrie(source map[string]HostAddress) NullAddressTrie {
+	return NullAddressTrie{
+		Trie:  NewAddressTrie(source),
+		Valid: true,
+	}
+}
+
+// MarshalJSON converts NullAddressTrie to valid JSON
+func (n NullAddressTrie) MarshalJSON() ([]byte, error) {
+	if !n.Valid { // envconfig bug?
+		return []byte("null"), nil
+	}
+
+	jsonMap := make(map[string]string)
+	for k, v := range n.Trie.source {
+		var right string
+		if v.Port != 0 {
+			right = v.String()
+		} else {
+			right = v.IP.String()
+		}
+		jsonMap[k] = right
+	}
+
+	return json.Marshal(jsonMap)
+}
+
+// UnmarshalJSON converts JSON to NullAddressTrie
+func (n *NullAddressTrie) UnmarshalJSON(data []byte) error {
+	if bytes.Equal(data, []byte("null")) {
+		n.Trie = nil
+		n.Valid = false
+		return nil
+	}
+
+	jsonSource := make(map[string]string)
+	if err := json.Unmarshal(data, &jsonSource); err != nil {
+		return err
+	}
+
+	source := make(map[string]HostAddress)
+	for k, v := range jsonSource {
+		ip, port, err := net.SplitHostPort(v)
+		if err == nil {
+			pInt, err := strconv.Atoi(port)
+			if err != nil {
+				return err
+			}
+
+			source[k] = HostAddress{IP: net.ParseIP(ip), Port: pInt}
+		} else {
+			source[k] = HostAddress{IP: net.ParseIP(v)}
+		}
+	}
+
+	n.Trie = NewAddressTrie(source)
+	n.Valid = true
+	return nil
+}
+
+// AddressTrie is wrapper around trieNode to integrate with net.TCPAddr
+type AddressTrie struct {
```

Comment:
Same comment as on the source `HostAddress` type. I would prefer to use the `Host` nomenclature instead of `Address`. Same comment for the overarching `NullAddressTrie` 🙇🏻

Reply:
Purging the `Address` suffix everywhere 🔫

Reply:
I agree with the `Hosts` over other names we mentioned.

Reply:
Also, I think, `BlockHostnames` is better than `Hostnames` which is very similar to `Hosts` also it's akin to the [k6 option name](https://k6.io/docs/using-k6/k6-options/reference/#block-hostnames). I can change that if you wish.

Reply:
I think the plural is enough of a distinguishing feature between `Host` and `Hosts`, so I'm OK with it.

I wouldn't rename `HostnameTrie` to `BlockHostnames` since there's nothing "blocking" about it. It could just as well be used as an allow list, so `Hostnames` is a generic term for the container. But I wouldn't rename it in this PR, let's leave it as `HostnameTrie` for now.